src/Utilities/Compiler/Extensions/IEnumerableExtensions.cs

```diff
@@ -103,6 +103,87 @@ public static void Dispose<T>(this IEnumerable<T> collection)
             }
         }
 
+        /// <summary>
+        /// Determines whether a sequence contains, exactly, <paramref name="count"/> elements.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">The <see cref="IEnumerable{TSource}"/> to check for cardinality.</param>
+        /// <param name="count">The number of elements to ensure exists.</param>
+        /// <returns><see langword="true" /> the source sequence contains, exactly, <paramref name="count"/> elements; otherwise, <see langword="false" />.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+        public static bool HasExactly<TSource>(this IEnumerable<TSource> source, int count)
```

Comment:
Have you benchmarked whether a type check for collections with O(1) size lookup (`ICollection`, `IReadOnlyCollection`, `Array`) would be faster? Doing so would avoid allocating enumerators. Ditto the other members on this class.

For example: https://github.com/dotnet/corefx/blob/ec138afba514ce9b53d931d66129e02e8be147a1/src/System.Linq/src/System/Linq/Count.cs#L19-L46

Reply:
That's right.

Reply:
Done!

Couldn't do the optimization for `IIListProvider<TSource>` because it's internal.